File: apple/Sources/BidiLens/UnicodeData.swift
import Foundation

public enum BidiUnicodeData {
    public static let version = GeneratedBidiRanges.unicodeVersion
    public static let bidiSHA256 = GeneratedBidiRanges.bidiSHA256
    public static let generalCategorySHA256 = GeneratedBidiRanges.generalCategorySHA256
}
enum UnicodeClassifier {
    static func contains(_ ranges: [UInt32], _ scalar: UInt32) -> Bool {
        var low = 0
        var high = ranges.count / 2 - 1
        while low <= high {
            let middle = (low + high) / 2
            let start = ranges[middle * 2]
            let end = ranges[middle * 2 + 1]
            if scalar < start {
                high = middle - 1
            } else if scalar > end {
                low = middle + 1
            } else {
                return true
            }
        }
        return false
    }

    static func classifyStrong(_ scalar: UInt32) -> BidiDirection {
        if contains(GeneratedBidiRanges.nonStrong, scalar) { return .neutral }
        return contains(GeneratedBidiRanges.rtl, scalar) ? .rightToLeft : .leftToRight
    }

    static func classifyNatural(_ scalar: UInt32) -> BidiDirection {
        contains(GeneratedBidiRanges.naturalLetters, scalar) ? classifyStrong(scalar) : .neutral
    }

    static func enumerate(_ text: String) -> [(scalar: UnicodeScalar, utf16: Int, codePoint: Int)] {
        var utf16 = 0
        var result: [(UnicodeScalar, Int, Int)] = []
        result.reserveCapacity(text.unicodeScalars.count)
        for (codePoint, scalar) in text.unicodeScalars.enumerated() {
            result.append((scalar, utf16, codePoint))
            utf16 += scalar.value > 0xffff ? 2 : 1
        }
        return result
    }

    static func codePointOffset(_ text: String, utf16Offset: Int) -> Int {
        var consumed = 0
        var count = 0
        for scalar in text.unicodeScalars {
            if consumed >= utf16Offset { break }
            consumed += scalar.value > 0xffff ? 2 : 1
            count += 1
        }
        return count
    }

    static func substring(_ text: String, utf16Range: Range<Int>) -> String {
        (text as NSString).substring(with: NSRange(
            location: utf16Range.lowerBound,
            length: utf16Range.count
        ))
    }
}

File: windows/src/BidiLens.Core/UnicodeData.cs
using System.Text;

namespace BidiLens;

public static class BidiUnicodeData
{
    public const string Version = GeneratedBidiRanges.UnicodeVersion;
    public const string BidiSha256 = GeneratedBidiRanges.BidiSha256;
    public const string GeneralCategorySha256 = GeneratedBidiRanges.GeneralCategorySha256;
}

internal static class UnicodeClassifier
{
    private static bool Contains(int[] ranges, int codePoint)
    {
        var low = 0;
        var high = ranges.Length / 2 - 1;
        while (low <= high)
        {
            var middle = (low + high) >>> 1;
            var start = ranges[middle * 2];
            var end = ranges[middle * 2 + 1];
            if (codePoint < start) high = middle - 1;
            else if (codePoint > end) low = middle + 1;
            else return true;
        }
        return false;
    }

    internal static BidiDirection ClassifyStrong(int value)
    {
        if (Contains(GeneratedBidiRanges.NonStrong, value)) return BidiDirection.Neutral;
        return Contains(GeneratedBidiRanges.Rtl, value)
            ? BidiDirection.RightToLeft
            : BidiDirection.LeftToRight;
    }

    internal static BidiDirection ClassifyNatural(int value) =>
        Contains(GeneratedBidiRanges.NaturalLetters, value)
            ? ClassifyStrong(value)
            : BidiDirection.Neutral;

    internal static bool IsCombiningMark(int value) =>
        Contains(GeneratedBidiRanges.CombiningMarks, value);

    internal static IEnumerable<(Rune Rune, int Utf16Index, int CodePointIndex)> Enumerate(string text)
    {
        var utf16 = 0;
        var codePoint = 0;
        foreach (var rune in text.EnumerateRunes())
        {
            yield return (rune, utf16, codePoint);
            utf16 += rune.Utf16SequenceLength;
            codePoint++;
        }
    }

    internal static int CodePointOffset(string text, int utf16Offset)
    {
        if (utf16Offset < 0 || utf16Offset > text.Length) throw new ArgumentOutOfRangeException(nameof(utf16Offset));
        var count = 0;
        var consumed = 0;
        foreach (var rune in text.EnumerateRunes())
        {
            if (consumed >= utf16Offset) break;
            consumed += rune.Utf16SequenceLength;
            count++;
        }
        return count;
    }
}
